### PR TITLE
[Fixes #2192] Remove enumerable types from excluded type in validation

### LINKED ISSUE #2192: `ModelStateDictionary` contains garbage entries for `DateTimeOffset[]`
This is a regression from Beta3.   I have not tested with other collecition types.

1. create a model that is a `DateTimeOffset[]` or that contains a property of that array type
2. create an action of the form `public IActionResult Dates([FromForm(Name = "")] DateTimeOffset[] model)` or whatever works for your more complex type
3. create a matching action that invokes a view submitting to above action
4. run in the debugger
5. hit Submit in the browser
6. break on the first line of the action from step 2
7. observe garbage in modelstate
8. note that even though the parameter is fully-populated and the model is valid, `ModelState.IsValid` is `false`.
9. a similar scenario using Beta3 (and `[Bind(Prefix ="")]` instead of `[FromForm(Name = "")]`) does not have `Length` or similar entries in the `ModelState`.  `ModelState.IsValid` is `true` when using that version.

/cc @harshgMSFT

### PR DESCRIPTION
**Pull request by ajaybhargavb**
@harshgMSFT @rynowak

### CODE REVIEW

**Comment by ajaybhargavb** [comment]
Updated with test.

File: src/Microsoft.AspNet.Mvc.ModelBinding/Validation/SimpleTypesExcludeFilter.cs [hunk 1]
```diff
@@ -19,22 +19,15 @@ public bool IsTypeExcluded(Type type)
         {
             Type[] actualTypes;
 
-            var enumerable = type.ExtractGenericInterface(typeof(IEnumerable<>));
-            if (enumerable == null)
+            // The following special case is for <KeyValuePair<K,V>
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Validation/SimpleTypesExcludeFilter.cs]: The comment should not say what is being done but why it is being done. Did you check what  happens for a dictionary?

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/SimpleTypeExcludeFilterTest.cs [hunk 2]
```diff
@@ -15,6 +15,10 @@ public void SimpleTypeExcluceFilter_ExcludedTypes(Type type)
         {
             // Arrange
             var filter = new SimpleTypesExcludeFilter();
+            if (type == typeof(int[]))
+            {
+                //System.Diagnostics.Debugger.Launch();
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/SimpleTypeExcludeFilterTest.cs]: remove please.

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Validation/DefaultObjectValidatorTests.cs [hunk 3]
```diff
@@ -472,19 +472,69 @@ public void NonRequestBoundModel_MarkedAsSkipped()
...
+            var validator = new DefaultObjectValidator(
+                testValidationContext.ExcludeFiltersProvider,
+                testValidationContext.ModelMetadataProvider);
+
+            // Act
+            validator.Validate(validationContext);
+
+            // Assert
+            Assert.True(validationContext.ModelState.IsValid);
+            var modelState = validationContext.ModelState["items"];
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Validation/DefaultObjectValidatorTests.cs]: Would there be an entry with items on the model state dictionary?

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Validation/DefaultObjectValidatorTests.cs [hunk 4]
```diff
@@ -472,19 +472,69 @@ public void NonRequestBoundModel_MarkedAsSkipped()
             validator.Validate(validationContext);
 
             // Assert
+            Assert.True(validationContext.ModelState.IsValid);
             var modelState = validationContext.ModelState["serviceProvider.TestService"];
             Assert.Empty(modelState.Errors);
             Assert.Equal(modelState.ValidationState, ModelValidationState.Skipped);
         }
 
+        [Fact]
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Validation/DefaultObjectValidatorTests.cs]: This should be a bigger set, should include all the types which you excluded from `SimpleTypeExcludeFilterTest`

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Validation/DefaultObjectValidatorTests.cs (hunk 3 above)
Comment (ajaybhargavb) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Validation/DefaultObjectValidatorTests.cs]: Yes. This https://github.com/aspnet/Mvc/blob/9eb874b77c253818c6754ab779fdf50e9e9a1475/src/Microsoft.AspNet.Mvc.ModelBinding/ModelStateDictionary.cs#L395 will create a new `ModelState` for the key `items`.

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/SimpleTypeExcludeFilterTest.cs (hunk 2 above)
Comment (ajaybhargavb) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/SimpleTypeExcludeFilterTest.cs]: Oops. Will remove.

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Validation/DefaultObjectValidatorTests.cs (hunk 3 above)
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Validation/DefaultObjectValidatorTests.cs]: ok this seems fine.

**Comment by ajaybhargavb** [comment]
Updated.

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/SimpleTypeExcludeFilterTest.cs [hunk 5]
```diff
@@ -43,24 +43,14 @@ public static TheoryData<Type> ExcludedTypes
...
-                    typeof(HashSet<string>),
-
-                    // Value types
-                    typeof(IList<DateTime>),
+                    typeof(int?),
 
                     // KeyValue types
-                    typeof(Dictionary<int, string>),
-                    typeof(IReadOnlyDictionary<int?, char>)
+                    typeof(KeyValuePair<string, string>)
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/SimpleTypeExcludeFilterTest.cs]: Ho about Key value pair of complex and non complex types?

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Validation/DefaultObjectValidatorTests.cs [hunk 6]
```diff
@@ -472,19 +472,132 @@ public void NonRequestBoundModel_MarkedAsSkipped()
...
+            // Arrange
+            var modelStateDictionary = new ModelStateDictionary();
+            modelStateDictionary.Add("items[0].Key", new ModelState());
+            modelStateDictionary.Add("items[0].Value", new ModelState());
+            modelStateDictionary.Add("items[1].Key", new ModelState());
+            modelStateDictionary.Add("items[1].Value", new ModelState());
+
+            var model = new Dictionary<string, string>()
+            {
+                {"FooKey", "FooValue" },
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Validation/DefaultObjectValidatorTests.cs]: spacing here and below.

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Validation/DefaultObjectValidatorTests.cs [hunk 7]
```diff
@@ -472,19 +472,132 @@ public void NonRequestBoundModel_MarkedAsSkipped()
...
+            };
+
+            var testValidationContext = GetModelValidationContext(
+                model,
+                typeof(Dictionary<string, string>),
+                "items",
+                excludedTypes: null,
+                modelStateDictionary: modelStateDictionary);
+
+            var excludedValidationTypesPredicate = new List<IExcludeTypeValidationFilter>();
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Validation/DefaultObjectValidatorTests.cs]: `excludeTypeFilters` ?

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Validation/DefaultObjectValidatorTests.cs [hunk 8]
```diff
@@ -472,19 +472,132 @@ public void NonRequestBoundModel_MarkedAsSkipped()
...
+                testValidationContext.ExcludeFiltersProvider,
+                testValidationContext.ModelMetadataProvider);
+
+            // Act
+            validator.Validate(validationContext);
+
+            // Assert
+            Assert.True(validationContext.ModelState.IsValid);
+            var modelState = validationContext.ModelState["items"];
+            Assert.Empty(modelState.Errors);
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Validation/DefaultObjectValidatorTests.cs]: no need for checking this if you are checking IsValid.

**Comment by harshgMSFT** [comment]
:watch:  for minor comments.

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/SimpleTypeExcludeFilterTest.cs (hunk 5 above)
Comment (ajaybhargavb) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/SimpleTypeExcludeFilterTest.cs]: That is included in TheoryData of IncludedTypes.

**Comment by ajaybhargavb** [comment]
Updated.

**Comment by harshgMSFT** [comment]
:shipit: from me.

**Pull request merged**